docs/content/guides/dagster/enriching-with-software-defined-assets.mdx

```diff
@@ -0,0 +1,312 @@
+```
+
+### A job where not all ops produce assets
+
+Here's a job where some of the ops - `extract_products` and `get_categories` don't producing an asset of its own - they just produce transient data that downstream ops will use to produce assets:
+
+```python file=/guides/dagster/enriching_with_software_defined_assets/vanilla_graph.py
+from pandas import DataFrame
+
+from dagster import job, op, repository
+
+from .mylib import create_db_connection, fetch_products
+
+
+@op
+def extract_products() -> DataFrame:
+    return fetch_products()
+
+
+@op
+def get_categories(products: DataFrame) -> DataFrame:
+    return DataFrame({"category": products["category"].unique()})
+
+
+@op
+def write_products_table(products: DataFrame) -> None:
+    products.to_sql(name="categories", con=create_db_connection())
+
+
+@op
+def write_categories_table(categories: DataFrame) -> None:
+    categories.to_sql(name="categories", con=create_db_connection())
+
+
+@job
+def ingest_products_and_categories():
+    products = extract_products()
+    product_categories = get_categories(products)
+    return write_products_table(products), write_categories_table(product_categories)
+
+
+@repository
+def repo():
+    return [ingest_products_and_categories]
+```
+
+Here's an equivalent job with software-defined assets. Note that, because not all the ops correspond to assets, it makes use of the `@op` and `@graph` APIs and uses `from_graph` to wrap a graph in a software-defined asset.
+
+```python file=/guides/dagster/enriching_with_software_defined_assets/sda_graph.py
+from pandas import DataFrame
+
+from dagster import AssetSelection, AssetsDefinition, GraphOut, graph, op, repository
+
+from .mylib import create_db_connection, fetch_products
+
+
+@op
+def extract_products() -> DataFrame:
+    return fetch_products()
+
+
+@op
+def get_categories(products: DataFrame) -> DataFrame:
+    return DataFrame({"category": products["category"].unique()})
+
+
+@op
+def write_products_table(products: DataFrame) -> None:
+    products.to_sql(name="categories", con=create_db_connection())
+
+
+@op
+def write_categories_table(categories: DataFrame) -> None:
+    categories.to_sql(name="categories", con=create_db_connection())
+
+
+@graph(out={"products": GraphOut(), "categories": GraphOut()})
+def ingest_graph():
+    products = extract_products()
+    product_categories = get_categories(products)
+    return write_products_table(products), write_categories_table(product_categories)
+
+
+two_tables = AssetsDefinition.from_graph(ingest_graph)
+
+
+@repository
+def repo():
+    return [two_tables, AssetSelection.all().to_job("products_and_categories")]
+```
+
+## Future work: jobs that run ops after assets
+
+In the future, you'll be able to define jobs that materialize software-defined assets and then run arbitrary ops afterwards. If this is something you think you might use, we'd love to hear from you in Slack or in a Github discussion.
+
+## Concept mapping
```

Comment:
This section is great, but it feels odd to have it at the end here. To me, that makes it feel sorta thrown in, when it actually contains some really useful info.

Perhaps we should move it up and re-name it - something like `What are the differences between ops, graphs, and software-defined assets?`

Reply:
The title "Concept mapping" isn't a great one.  What I'm trying to get across here isn't about the differences though. The way I imagined it fits into the overall narrative is that the earlier parts of the blog post give the broad strokes of how to migrate in common cases.  This is covering the long tail.  The aim is to answer questions like "If I start using software-defined assets, does that mean I no longer get to use config?"

Reply:
Gotcha. I'll think on this a bit.